### DISCUSSION TITLE
Interactive Fuzzy-Filter or Keyboard-Driven Prompt

### DISCUSSION THREAD

**Discussion by KyleKing** [Textualize/rich#2453] [Q&A] [answered]
I'm new to Rich and I wondered if there was a way to create an interactive prompt. Either keyboard-driven or fuzzy-matching?

---

I saw https://github.com/Textualize/rich/pull/2192 where fuzzy-filtering is implemented (which I'm using for now), but I would like to figure out a way to live-filter the choices

I'm looking for something like [`gum`'s](https://github.com/charmbracelet/gum) `gum choose "fix" "feat" "docs" "style" "refactor" "test" "chore" "revert"`

![gum](https://camo.githubusercontent.com/e55a3f3371335e08446eac4192c9749f6f01eae9f30654360774e29b26705331/68747470733a2f2f73747566662e636861726d2e73682f67756d2f636f6d6d69745f322e676966)

**Comment by KyleKing** [comment, marked as answer]
TLDR: not supported in rich (https://github.com/Textualize/rich/issues/392#issuecomment-711873424 & https://github.com/Textualize/rich/issues/266#issuecomment-687642551), but possible with textual (https://github.com/Textualize/rich/issues/1843#issuecomment-1018714294) or could be done with [`questionary`](https://github.com/tmbo/questionary). This is partially related: https://github.com/Textualize/rich/discussions/960

However, after looking at a bunch of the rich example apps, it seems like a common workaround is to provide indexing, so you could do something like:

```python
choices = ["fix","feat","docs"]

for idx, choice in enumerate(choices):
    console.print(f"{idx}. {choice}")
selection = Prompt.ask('Which option?', choices=[*map(str, range(len(choices)))], default="0")
```

Look like:

```sh
0. fix
1. feat
2. docs
Which option? [0/1/2] (0):
```

**Comment by FergusFettes** [comment]
With iterfzf you can drop into a terminal and use fzf there quite easily, I'm integrating it with rich just now and it works well:
```
    import iterfzf
    choices = ["apple", "banana", "cherry", "date", "elderberry"]
    selected = iterfzf.iterfzf(choices)
    print(selected)
```

**Reply by KyleKing** [reply]
That's a good find! Unfortunately looks like there hasn't been a recent release though and it pins a very old version of `fzf`: https://github.com/dahlia/iterfzf/blob/d030d71a6748b8507735458e0acadcddc7a6cd01/setup.py#L23-L26

**Reply by FergusFettes** [reply]
tbh, just using subprocess with fzf also works pretty well :D

```
import subprocess
choices = ["apple", "banana", "cherry", "date", "elderberry"]
p = subprocess.Popen(["fzf"], stdin=subprocess.PIPE, stdout=subprocess.PIPE)
output, _ = p.communicate("\n".join(choices).encode())
print(output.decode())
```
